doc/dev/kdb-record.md

```diff
@@ -0,0 +1,110 @@
+# `kdb record` Command
+
+The `kdb record` tooling enables the user to record configuration sessions. It also allows exporting and importing the configuration changes.
+
+Recording sessions are stored on a per-user basis. This implies that a recording session can only track changes made by a single user.
+
+## Session handling
+
+* `kdb record start [session_name]`: Starts a new recording session. If no session name is provided, a new random name is generated. If a session with the provided name already exists, changes will be appended.
+* `kdb record stop`: Stops the current recording session. 
+* `kdb record delete [session_name]`: Deletes the specified recording session. If no session name is provided, the current session will be deleted. Deleting a session will not stop the recording. 
+* `kdb record list-sessions`: Lists all sessions by their name
+* `kdb record changes [session_name]`: Shows all changes for the specified session. If no session name is provided, the current session will be used. 
+
+## Exporting configuration changes
+
+`kdb record export [session_name] [--format <format>]`: Will dump the changed keys of the specified session onto `stdout`. If no session name is provided, the current session will be used. 
+
+As example, suppose the key `user:/my/app/hello` has the value `world` and you execute the following commands:
+* `kdb set user:/my/app/hello max`
+* `kdb set user:/my/app/hello susi`
+
+Only the value `susi` will be exported. Additionally, if you execute `kdb set user:/my/app/hello world` the key will not be exported at all, as there are no changes compared to the beginning of the recording.
+
+The same principle applies to the creation and deletion of keys within a session. If you create a key and delete it in the same session, it will not be exported.
+
+By default, the elektra dump format is used. The optional --format parameter can be used to specify which storage plugin should be used to format the output.
```

Comment:
Yes, as the storage plugins do not support removing keys, we have to come up with a different format. I'm currently thinking that something (loosely) based on the standard .patch file format could work.

As I do not want to hard code this format in the exporter, we may have to come up with new "export" and "import" functions that plugins can implement. This would enable having multiple file formats for export in the future.

Reply:
Please also look at the merging strategies. With merging strategies you don't need to represent "removed keys", as all keys present in base but not present in the result are removed. Please also update doc/help/elektra-merge-strategy.md as you think it should be for the new tooling.

A representation for multiple KeySets, however, will be needed, because this is also needed for having several namespaces in one import/export file.

Reply:
Please also look at how merging works :wink: 

The base doesn't contain user:/foo (assuming `kdb record` defines the base), ours doesn't contain user:/foo, so theirs wins without conflict: user:/foo will stay present with `456`.

Reply:
How would you now what the base is here? AFAICT this would only be possible, if `kdb record start` makes a copy of the whole KDB (or below some parent key given as an argument). And if we do that, we don't need to "merge recorded changes into a diff". Instead we can just produce a diff against the base during `kdb record stop`.